cluster-dns (Part 2/2)

thz opened the pull request:
**What this PR does / why we need it**:

This PR adds openvpn sidecars for `scheduler` and `controllermanager` and configures master components' pods to use the user-cluster's DNS resolver.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #871 

**Special notes for your reviewer**:


Release note:
```release-note
Master components can now talk to cluster DNS
```

thz committed:
Add more OpenVPN sidecars and use user cluster DNS

thz commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -90,6 +85,23 @@ func Deployment(data *resources.TemplateData, existing *appsv1.Deployment) (*app
+	dep.Spec.Template.Spec.DNSConfig = &corev1.PodDNSConfig{
+		Nameservers: []string{data.UserClusterDNSResolverIP()},
+		Searches: []string{
+			"kube-system.svc.cluster.local",
```
Not sure if master components should have a search list at all.

mrIncompetent replied:
Is it required? I would like to omit as much as possible.
Though if kubernetes defaults those fields to `kube-system.svc.cluster.local` & `svc.cluster.local` then we should write those here as well - otherwise our controller detects a difference between desired & actual state -> resulting in an endless loop of updates

thz replied:
I prefer an empty search list. Of course this will break uses like "apiserver" without any suffix. But perhaps there is no such use less after Part1/2 of the cluster-dns proposal implementation.

I will check for such a cycle. Originally k8s creates more entries:
```
nameserver 10.47.240.10
search cluster-4b8zj9pvch.svc.cluster.local svc.cluster.local cluster.local c.kubermatic-dev.internal google.internal
options ndots:5
```

This is mixed with what comes from the host/node itself.
But with `DNSPolicy=None` it does not mix in the hosts search list.
I will do some more testing.

thz committed:
Fixtures.

(This time it was vim macro on test-output piped to /bin/patch.
 Next time it will be a script extracting the patch from
 test failure output...)

thz commented on api/pkg/resources/resources.go:
```diff
@@ -223,6 +223,23 @@ func (d *TemplateData) ClusterIPByServiceName(name string) (string, error) {
 	return service.Spec.ClusterIP, nil
 }
 
+// UserClusterDNSResolverIP returns the 9th usable IP address
```
Just temporary in this PR. This might get obsoleted by the outcome of #1357

mrIncompetent replied:
This actually implements parts of #1357
If this mirrors the same logic like the addon-controller i'm fine with the duplication for now

thz replied:
Yes - it is the same logic as in #1357 `getKubeDNSClusterIP()`

thz commented:
PTAL @mrIncompetent

mrIncompetent commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -106,65 +118,12 @@ func Deployment(data *resources.TemplateData, existing *appsv1.Deployment) (*app
 	}
 
-	openvpnServiceIP, err := data.ClusterIPByServiceName(resources.OpenVPNServerServiceName)
+	openvpnSidecar, err := resources.OpenVPNSidecarContainer(data, "openvpn-client")
```
:+1:

mrIncompetent commented on api/pkg/resources/apiserver/deployment.go:
```diff
@@ -90,6 +85,23 @@ func Deployment(data *resources.TemplateData, existing *appsv1.Deployment) (*app
+		Options: []corev1.PodDNSConfigOption{
+			{
+				Name:  "ndots",
+				Value: &dnsConfigOptionNdots,
```
Whats this for?

thz replied:
What libc does on resolving is: trying to resolve as a fqdn and then adding all entries from search list and try to resolve that.
`ndots` is the minimum number of dots which is required to check if the given name resolves as a fqdn without any suffix. Default (libc) is 1 IIRC, but because of heavy use of subdomains in k8s that got increased.

thz committed:
No DNS searches for user-cluster master components

* The master components (apiserver,scheduler,controllermanager) of
  the user cluster now get an empty dns-resolver search list.
* Moved common code for Pod DNS config into pkg/resources.

thz committed:
Fixtures.

mrIncompetent commented:
@thz Did you tested this PR according to https://github.com/kubermatic/kubermatic/blob/master/docs/proposals/cluster-dns.md#task--effort ?

thz committed:
Make UserClusterDNSResolverIP use the Cluster

Changes the signature from UserClusterDNSResolverIP from:
TemplateData.UserClusterDNSResolverIP() string
to:
UserClusterDNSResolverIP(*Cluster) string
which allows using it without TemplateData which was
only used to get the Cluster.

thz committed:
Fixtures for DNSConfig change.

thz committed:
Add cluster.local to search list.

thz committed:
Add admission-webhook test files.

mrIncompetent commented:
@thz Whats left on this PR?

thz commented:
Just the e2e run is left. @mrIncompetent

mrIncompetent commented on api/pkg/resources/resources.go:
```diff
@@ -223,6 +223,43 @@ func (d *TemplateData) ClusterIPByServiceName(name string) (string, error) {
+	return corev1.DNSNone, &corev1.PodDNSConfig{
+		Nameservers: []string{UserClusterDNSResolverIP(cluster)},
+		Searches: []string{
+			"cluster.local",
```
`cluster.Spec.ClusterNetwork.DNSDomain` instead if hardcoded `cluster.local`

thz committed:
Adapt for API changes after rebase...

thz committed:
Fix: do not hardcode "cluster.local"

alvaroaleman commented on api/pkg/resources/resources.go:
```diff
@@ -247,6 +247,33 @@ func UserClusterDNSResolverIP(cluster *kubermaticv1.Cluster) (string, error) {
+	return corev1.DNSNone, &corev1.PodDNSConfig{
+		Nameservers: []string{dnsConfigResolverIP},
+		Searches: []string{
+			cluster.Spec.ClusterNetwork.DNSDomain,
```
What about `${NS}.svc.${DNSDOMAIN]` and `svc.${DNSDOMAIN}`?

thz replied:
See discussion in this PR. Current Idea is to go with the shortest possible list. I have no strong opinion here, so perhaps we need to reopen that discussion?

alvaroaleman replied:
I would prefer to stay with the kube defaults here, just to ensure we don't end up with some weird issues that are ultimatively caused by the apiserver not having the correct search domains

thz replied:
@mrIncompetent, opinion?

mrIncompetent replied:
Nope

thz replied:
@alvaroaleman I added back the `kube-system.svc.${DNSDOMAIN}` and `svc.${DNSDOMAIN}`.

alvaroaleman commented on api/cmd/image-loader/main.go:
```diff
@@ -293,6 +293,7 @@ func getTemplateData(versions []*version.MasterVersion, requestedVersion string)
 	fakeCluster.Spec.Version = masterVersion.Version.String()
 	fakeCluster.Spec.ClusterNetwork.Pods.CIDRBlocks = []string{"172.25.0.0/16"}
 	fakeCluster.Spec.ClusterNetwork.Services.CIDRBlocks = []string{"10.10.10.0/24"}
+	fakeCluster.Spec.ClusterNetwork.DNSDomain = "cluster.local"
```
Is this required?

thz replied:
I prefer to return an error in `UserClusterDNSPolicyAndConfig` when the DNSDomain is empty. Unfortunately the image-loader fails then, when fakeCluster has no domain.

alvaroaleman commented on config/test/admission-webhooks/admissionvalidator-svc.yaml:
```diff
@@ -0,0 +1,122 @@
+apiVersion: apps/v1
```
Any chance we can actually deploy and test this with the existing e2e runner? Doesn't have to be in this pull, but maybe just create a ticket for it

thz replied:
The problem is that the e2e runner fails even without this change.
Currently this is tested with admission webhook and reachability of an extension apiserver.

thz committed:
Add defaults back to dns search list.

thz committed:
Update fixtures.

thz commented:
```
2018-07-05T14:07:07+0200 Ran 154 of 891 Specs in 392.394 seconds
2018-07-05T14:07:07+0200 SUCCESS! -- 154 Passed | 0 Failed | 0 Pending | 737 Skipped 
```
e2e conformance test passed.

thz commented:
Meanwhile a new commit landed and I had to rebase...
